blockchain/chain.go

```diff
@@ -86,27 +86,38 @@ type orphanBlock struct {
 // However, the returned snapshot must be treated as immutable since it is
 // shared by all callers.
 type BestState struct {
-	Hash         chainhash.Hash // The hash of the block.
-	Height       int64          // The height of the block.
-	Bits         uint32         // The difficulty bits of the block.
-	BlockSize    uint64         // The size of the block.
-	NumTxns      uint64         // The number of txns in the block.
-	TotalTxns    uint64         // The total number of txns in the chain.
-	MedianTime   time.Time      // Median time as per CalcPastMedianTime.
-	TotalSubsidy int64          // The total subsidy for the chain.
+	Hash           chainhash.Hash   // The hash of the block.
+	PrevHash       chainhash.Hash   // The previous block hash.
+	Height         int64            // The height of the block.
+	Bits           uint32           // The difficulty bits of the block.
+	PoolSize       uint32           // The ticket pool size.
+	NextStakeDiff  int64            // The next stake difficulty.
+	BlockSize      uint64           // The size of the block.
+	NumTxns        uint64           // The number of txns in the block.
+	TotalTxns      uint64           // The total number of txns in the chain.
+	MedianTime     time.Time        // Median time as per CalcPastMedianTime.
+	TotalSubsidy   int64            // The total subsidy for the chain.
+	WinningTickets []chainhash.Hash // The eligigle tickets to vote on the next block.
+	MissedTickets  []chainhash.Hash // The missed tickets set to be revoked.
+	FinalState     [6]byte          // The calculated state of the lottery.
 }
 
 // newBestState returns a new best stats instance for the given parameters.
-func newBestState(node *blockNode, blockSize, numTxns, totalTxns uint64, medianTime time.Time, totalSubsidy int64) *BestState {
+func newBestState(node *blockNode, blockSize, numTxns, totalTxns uint64, medianTime time.Time, totalSubsidy int64, nextStakeDiff int64) *BestState {
 	return &BestState{
-		Hash:         node.hash,
-		Height:       node.height,
-		Bits:         node.bits,
-		BlockSize:    blockSize,
-		NumTxns:      numTxns,
-		TotalTxns:    totalTxns,
-		MedianTime:   medianTime,
-		TotalSubsidy: totalSubsidy,
+		Hash:           node.hash,
+		PrevHash:       node.parent.hash,
+		Height:         node.height,
+		Bits:           node.bits,
+		PoolSize:       node.poolSize,
+		NextStakeDiff:  nextStakeDiff,
+		BlockSize:      blockSize,
+		NumTxns:        numTxns,
+		TotalTxns:      totalTxns,
+		MedianTime:     medianTime,
+		TotalSubsidy:   totalSubsidy,
+		WinningTickets: node.stakeNode.Winners(),
```

Comment:
This could easily crash with `node.stakeNode` not being loaded since the field are dynamically loaded.  You need to make sure that you call `fetchStakeNode` prior to invoking `newBestState` with these changes.

In the case of `connectBlock` and `disconnectBlock`, there are already calls to load the stake nodes, but they are after the calls to `newBestState`, so the loading code should be moved above the blocks of code which deal with setting the best state.

In the case of the init path, you'll need to load the stake nodes.